seqr/views/apis/anvil_workspace_api_tests.py

```diff
@@ -344,19 +385,40 @@ def test_validate_anvil_vcf(self, mock_file_iter, mock_file_exist, mock_utils_lo
         self.assertEqual(response.json()['error'],
                          'Invalid VCF file format - file path must end with .vcf or .vcf.gz or .vcf.bgz')
 
+        # test no header line
         mock_file_exist.return_value = True
-        mock_file_iter.return_value = ['##fileformat=VCFv4.2\n',
-                                       '#CHROM	POS	ID	REF	ALT	QUAL']  # incomplete header line
+        mock_file_iter.return_value = BASIC_META + DATA_LINES
         response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY_GZ_DATA_PATH))
         self.assertEqual(response.status_code, 400)
-        self.assertEqual(response.json()['error'],
-                         'No samples found in the provided VCF. This may be due to a malformed file')
+        self.assertListEqual(response.json()['errors'], [
+            'No header found in the VCF file.'
+        ])
         mock_file_iter.assert_called_with('gs://test_bucket/test_path.vcf.gz', byte_range=(0, 65536))
         mock_file_exist.assert_called_with('gs://test_bucket/test_path.vcf.gz', user=self.manager_user)
 
+        # test header errors
+        mock_file_iter.return_value = BASIC_META + BAD_INFO_META + BAD_FORMAT_META + BAD_HEADER_LINE + DATA_LINES
+        response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY_GZ_DATA_PATH))
+        self.assertEqual(response.status_code, 400)
+        self.assertListEqual(response.json()['errors'], [
+            'No samples found in the provided VCF.',
+            'Missing required VCF header field(s) POS, FILTER, INFO, FORMAT.'
```

Comment:
so as per my previous review, when the header is invalid it means the file is totally malformed and we should do no other validation. You need to do 3 distinct steps of validation: validate the header, validate the samples, validate the metadata. If any of those validations fail you dont continue on to the next stage of validation and you only return the error s for that validation step, you do not run more validation and combine the errors

Reply:
I agree with the 3-step validation. We can extract samples after the header is validated.